### DISCUSSION TITLE
Fine tuning x-effect usage with multiple x-effect directives?

### DISCUSSION THREAD

**Discussion by trych** [alpinejs/alpine#3958] [1. Help] [answered]
Hi there!

I have a question about using `x-effect` in more complex setups, where a few properties need to be observed, but not all necessarily depend on the same events, which means that some should be re-evaluated at certain events while others should be re-evaluated at certain other events.

To not re-evaluate them all every time, I tried to split them up into several `x-effect` statements via appending a "pseudo-modifier" to `x-effect`, something like `x-effect.foo` to allow for more than one `x-effect` directive on one DOM node.

Here is some relevant excerpts of my specific example (the whole thing is supposed to be an image full screen view component with zoom and pan capabilities):

```html
<picture
  x-effect.displayedImg="console.log('displayedImg');
    if(aspect < viewportAspect) {
      displayedImgWidth = ww;
      displayedImgHeight = ww / aspect;
    } else {
      displayedImgWidth = wh * aspect;
      displayedImgHeight = wh;
    }
  "

  x-effect.maxTranslate="console.log('maxTranslate');
    maxTranslateX = (scale * displayedImgWidth - ww) / 2;
    maxTranslateY = (scale * displayedImgHeight - wh) / 2
  "

  x-effect.xy="console.log('xy');
    x = map(mouseX, 0, ww, maxTranslateX, -maxTranslateX);
    y = map(mouseY, 0, wh, maxTranslateY, -maxTranslateY);
    console.log(x);
  "
 >
  <img
    :style="{
       transform: `translate(-50%, -50%) translate(${x}px, ${y}px) scale(${scale})`
    }"
  >
 </picture>
``` 

(to make more sense of the code, the [full demo would be here](https://trych.dev/sandbox/fullscreen-img/index_v1.html))

Now, generally this works, this setup updates only the properties that should be updated according to certain events, for example `x` and `y` will only be updated when the mouse is moved, while the displayedImg stays untouched etc. Still, I have a few questions and one remaining issue.

1. This *feels* like a hack. Is this a pattern that makes sense or will I break anything this way?
2. My general implementation of the fullscreen component aside (it's work in progress), is there another better pattern that I could / should use instead? I tried `$watch`, but I would need to watch a *lot* of properties, it's not initially evaluated and it is not cleaned up after removal, so I would prefer `x-effect`.
3. The console statements suggest that all reactive changes "propagate through" (viewportAspect is changed, which updates displayedImgWidth, which updates maxTranslateX, which updates x, etc.) and it will also log the correct new value of `x` when changing the scale for instance. However, in this case, it does *not* update the transform value in the bound style of the img. This update would only work correctly when I include `scale` directly in the `x-effect.xy` like so:

```html
  x-effect.xy="scale;
    x = map(mouseX, 0, ww, maxTranslateX, -maxTranslateX);
    y = map(mouseY, 0, wh, maxTranslateY, -maxTranslateY);
    console.log(x);
  "
  ```
  
I do not quite understand why this correctly logs, but does not update the binding when scale is missing. To observe the error, you could try the demo, show the image, click to zoom in, click again to zoom out: now the position of the image is not correctly updated although the correct x value logs. Any idea, why this does not work?

Thanks a lot!

**Comment by trych** [comment]
Also, for easier reference, here is the full code of the demo (as I said WIP, so please don't get too concerned about the implementation of the component itself :) ).

```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>🖼 Fullscreen Img</title>

  <script src='../js/vendor/alpinejs@latest.min.js' defer></script>

  <link rel="stylesheet" type="text/css" href="./styles.css">

</head>
<body
  x-data="{
    fullscreen: false,
    fullscreenInitialized: false,

    ww: window.innerWidth,
    wh: window.innerHeight,
    viewportAspect: 1
  }"

  x-effect="viewportAspect = ww / wh"

  @resize.window.debounce="ww = window.innerWidth; wh = window.innerHeight;console.log('resize');"
  @keyup.esc.window="fullscreen && (fullscreen = false)"
>

  <button
    @click="fullscreenInitialized = true; fullscreen = !fullscreen"
    @mouseenter="fullscreenInitialized = true"
    x-text="fullscreen ? 'Hide Image' : 'Show Image'"
  >Show Image</button>

  <div
    class="fullscreen"

    x-show="fullscreen"
    x-cloak
  >

    <picture
      x-data="{
        baseUrl: '../assets/responsive_images/witz_',
        sizes: [420, 768, 1024, 1440, 1920, 2400],
        sources: $el.querySelectorAll('source'),

        srcset(format) {
          return this.sizes.map(size => `${this.baseUrl}${size}x.${format} ${size}w`).join(',');
        },

        aspect: 1,
        scale: 1,
        x: 0,
        y: 0,
        mouseX: ww / 2,
        mouseY: wh / 2,
        displayedImgWidth: ww,
        displayedImgHeight: wh,
        maxTranslateX: 0,
        maxTranslateY: 0,

        moveCursor(evt) {
          this.mouseX = evt.touches ? evt.touches[0].clientX : evt.clientX;
          this.mouseY = evt.touches ? evt.touches[0].clientY : evt.clientY;
        },

        map(i, minIn, maxIn, minOut, maxOut) {
          return (i - minIn) / (maxIn - minIn) * (maxOut - minOut) + minOut;
        }

      }"

      x-effect.displayedImg="console.log('displayedImg');
        if(aspect < viewportAspect) {
          displayedImgWidth = ww;
          displayedImgHeight = ww / aspect;
        } else {
          displayedImgWidth = wh * aspect;
          displayedImgHeight = wh;
        }
      "

      x-effect.maxTranslate="console.log('maxTranslate');
        maxTranslateX = (scale * displayedImgWidth - ww) / 2;
        maxTranslateY = (scale * displayedImgHeight - wh) / 2
      "

      x-effect.xy="console.log('xy');
        x = map(mouseX, 0, ww, maxTranslateX, -maxTranslateX);
        y = map(mouseY, 0, wh, maxTranslateY, -maxTranslateY)
      "

      @resize.window.debounce="if(fullscreenInitialized) {
        // see https://github.com/ausi/respimagelint/issues/31#issuecomment-420441005
        sources.forEach(s => {s.sizes += ''});
      }"

      @mousemove="moveCursor($event)"
      @touchmove="moveCursor($event)"
    >
      <source
          :srcset="fullscreenInitialized && srcset('avif')"
          sizes="100vw"
          type="image/avif">
      <source
          :srcset="fullscreenInitialized && srcset('webp')"
          sizes="100vw"
          type="image/webp">
      <source
          :srcset="fullscreenInitialized && srcset('jpg')"
          sizes="100vw"
          type="image/jpeg">
      <img
        data-src="../assets/responsive_images/witz_1024x.jpg"
        alt="Gemälde der Ratschluß der Erlösung von Konrad Witz"

        :src="fullscreenInitialized && (baseUrl + '1024x.jpg')"
        :style="{
          transform: `translate(-50%, -50%) translate(${x}px, ${y}px) scale(${scale})`
        }"

        @click="scale = (scale === 2) ? 1 : 2"
        @load.once="aspect = $el.naturalWidth / $el.naturalHeight; console.log('load')"
      >
    </picture>
  </div>

</body>
</html>
```

**Comment by ekwoka** [comment, marked as answer]
You can use Alpine.effect in the js side of things, like in the data's `init` method.

You might also want to look into the fullscreen api, instead of doing it this way.

**Reply by trych** [reply]
Thanks, this is indeed a bit cleaner:

```html
<picture
  x-init="
    Alpine.effect(() => {
      if(aspect < viewportAspect) {
        displayedImgWidth = ww;
        displayedImgHeight = ww / aspect;
      } else {
        displayedImgWidth = wh * aspect;
        displayedImgHeight = wh;
      }
    });

    Alpine.effect(() => {
      maxTranslateX = (scale * displayedImgWidth - ww) / 2;
      maxTranslateY = (scale * displayedImgHeight - wh) / 2;
    });

    Alpine.effect(() => {
      scale; // need this to properly update the bound values, when manually changing scale
             // although the updating of maxTranslateX actually triggers this effect ... ? 
      x = map(mouseX, 0, ww, maxTranslateX, -maxTranslateX);
      y = map(mouseY, 0, wh, maxTranslateY, -maxTranslateY);
    });
  "
></picture>
```

However, there is still the bug that I need to put the `scale` property directly into the same effect function to properly update the binding of `x` and `y` when the chain is triggered by a window resize for example. The effect seems to run (according to log statements) and sets the x value correctly, but does not update the bound value.

**Reply by SimoTod** [reply]
Did you check the element with the dom inspector? More likely the value is updated but the browser is not repainting it

**Reply by trych** [reply]
Yes, I have checked. The DOM inspector shows that the value is _not_ being updated, although the log statement suggests that the effect did run.

**Reply by ekwoka** [reply]
Hmm, can you recreate on codepen?

It SHOULD work, and it not working would be a bug. But there might be a way to track it down.

**Reply by trych** [reply]
Yes, here's a codepen: https://codepen.io/trych/pen/yLwJJpa

The example is a bit random, but it demonstrates that changing the `scale` value _does_ trigger the consequent effects (that changes the `x` value) and updates the reactive data accordingly (as seen by the `x-text` outputs), but the style binding does not get updated. Instead it gets incorrectly updated on the next toggle, so the x position of the cirlcle will always be wrong and not match the reactive data.

**Reply by ekwoka** [reply]
Okay, that's a good demonstration.

<img width="235" alt="Screenshot 2024-01-09 at 2 06 04 PM" src="https://github.com/alpinejs/alpine/assets/43540491/45c72a21-60e9-46a3-86f7-b5cc7c7bed71">

It definitely has to do with the scheduler trying to prevent over-running the effect.

I think I fixed this in a local experimental build and will check if so and open a PR.

**Reply by trych** [reply]
Great! "Glad" to hear that it was indeed something wrong with Alpine.js instead of me making some random rookie mistake. It was a bit hard to figure out within my more complex setup.
